Hex bin digit grouping

cgm616 opened the pull request:
This revives and updates an old pr (#3391) for the current API. 

Closes #2538.

---

*Please keep the line below*
changelog: Add [`unusual_byte_groupings`] lint.

rust-highfive commented:
r? @flip1995

(rust_highfive has picked a reviewer for you, use r? to override)

cgm616 commented:
I think this PR is all set now. There was one change made with the suggested grouping for hexadecimal literals that might need discussion. Instead of groups of 4, like `0xDEAD_BEEF`, clippy will now suggest groups of two, like `0xDE_AD_BE_EF`. This is to make sure that clippy's fixes won't contravene this lint, which suggests byte-sized chunks for hex literals.

flip1995 commented on clippy_lints/src/literal_representation.rs:
```diff
@@ -82,6 +82,25 @@ declare_clippy_lint! {
 }
 
+declare_clippy_lint! {
+    /// **What it does:** Warns if a hexadecimal or binary literal is not grouped
```
```suggestion
    /// **What it does:** Warns if a hexadecimal or binary literals are not grouped
```

flip1995 reviewed (commented):
I really don't like the suggestion change of the other lint. IMO splitting hex numbers in groups of 4 is more accepted than groups of 2.

cgm616 commented:
> I really don't like the suggestion change of the other lint. IMO splitting hex numbers in groups of 4 is more accepted than groups of 2.

Yeah, that's fair. I just reverted that change and switched this new lint to checking if hex literals are in groups of four to stay consistent.

flip1995 commented on tests/ui/literals.rs:
```diff
@@ -7,10 +7,10 @@
-    let ok4 = 0xab_cd_i32;
-    let ok5 = 0xAB_CD_u32;
-    let ok5 = 0xAB_CD_isize;
+    let ok3 = 0xabcd;
+    let ok4 = 0xabcd_i32;
+    let ok5 = 0xABCD_u32;
+    let ok5 = 0xABCD_isize;
```
We should not lint for groupings of 2 in hex literals either.

flip1995 commented on tests/ui/literals.stderr:
```diff
@@ -69,5 +69,25 @@ error: digits grouped inconsistently by underscores
+error: digits of hex or binary literal not grouped by four
+  --> $DIR/literals.rs:39:16
+   |
+LL |     let ok26 = 0x6_A0_BF;
```
This is definitely a FP

cgm616 replied:
Sorry, FP?

ebroto replied:
FP is an abbreviation of false positive

cgm616 replied:
I think I have fixed that, now.

cgm616 commented:
For some reason, my local builds always fail with seven or so of this error: 

```
no variant or associated item named `ConstBlock` found for enum `rustc_ast::ExprKind`
```

Does anyone know why? It makes debugging this lint really hard since I can't run any of the tests.

ebroto commented:
>my local builds always fail with seven or so of this error

I think you need to run `setup-toolchain.sh` again to catch up with the supported rustc version. See `doc/basics.md`.

cgm616 commented:
@rustbot modify labels: +S-waiting-on-review -S-waiting-on-author

cgm616 commented:
> > my local builds always fail with seven or so of this error
> 
> I think you need to run `setup-toolchain.sh` again to catch up with the supported rustc version. See `doc/basics.md`.

That helps partially; thank you. Now, though, I have a bunch of errors about crates being built with incompatible versions of `rustc`. I've tried deleting the `target/` folder and running `cargo clean`, but neither work.

ebroto commented:
>I have a bunch of errors about crates being built with incompatible versions of rustc

Could you share the errors?

Also, make sure you rebased on top of the current master.

cgm616 commented:
> Could you share the errors?

They went away mysteriously after waiting a while. It's possible that rust-analyzer running in VSCode had something to do with it. Thanks for the help!

flip1995 commented:
> It's possible that rust-analyzer running in VSCode had something to do with it.

I have the `    "rust-analyzer.checkOnSave.extraArgs": ["--target-dir", "target/rust-analyzer"],` config option set for rust-analyzer (I use coc.nvim, but I'm sure there's something similar for VSCode). This will create a `rust-analyzer` subdir in the target dir (for every project), so no more conflicting `cargo build` and `rust-analyzer` builds. I only suggest to use that if you have enough memory to spare.

flip1995 commented on tests/ui/literals.rs:
```diff
@@ -7,10 +7,10 @@
-    let ok4 = 0xab_cd_i32;
-    let ok5 = 0xAB_CD_u32;
-    let ok5 = 0xAB_CD_isize;
+    let ok3 = 0xabcd;
+    let ok4 = 0xabcd_i32;
+    let ok5 = 0xABCD_u32;
+    let ok5 = 0xABCD_isize;
```
These changes to the test file should be unnecessary now?

cgm616 replied:
Yep!

flip1995 commented on clippy_lints/src/literal_representation.rs:
````diff
@@ -82,6 +82,25 @@ declare_clippy_lint! {
+    /// let x: u32 = 0xFFF_FFF;
+    /// let y: u8 = 0b01_011_101;
+    /// ```
+    pub UNUSUAL_BYTE_GROUPING,
````
```suggestion
    pub UNUSUAL_BYTE_GROUPINGS,
```

Make sure to run `cargo dev update_lints` and to update the reference files after this change.

flip1995 reviewed (commented):
Two small things left, before this is ready to get merged.

flip1995 commented:
Thanks! Waiting on rustup.

@cgm616 I removed the `WIP` from the title. Was there something you thought was still missing?

cgm616 commented:
> @cgm616 I removed the `WIP` from the title. Was there something you thought was still missing?

Nope! Just leftover from the beginning. Thanks for everything!

flip1995 commented:
@bors r+

Thanks!

bors commented:
:pushpin: Commit 312bbff6968dbebd367ca90677c676e2cf5198d2 has been approved by `flip1995`

<!-- @bors r=flip1995 312bbff6968dbebd367ca90677c676e2cf5198d2 -->
<!-- homu: {"type":"Approved","sha":"312bbff6968dbebd367ca90677c676e2cf5198d2","approver":"flip1995"} -->

bors commented:
:hourglass: Testing commit 312bbff6968dbebd367ca90677c676e2cf5198d2 with merge eceebc344887269a821d1df0c8693d9eb4d5da7d...
<!-- homu: {"type":"BuildStarted","head_sha":"312bbff6968dbebd367ca90677c676e2cf5198d2","merge_sha":"eceebc344887269a821d1df0c8693d9eb4d5da7d"} -->

bors commented:
:sunny: Test successful - [checks-action_dev_test](https://github.com/rust-lang/rust-clippy/runs/1305369781), [checks-action_remark_test](https://github.com/rust-lang/rust-clippy/runs/1305367820), [checks-action_test](https://github.com/rust-lang/rust-clippy/runs/1305406490)
Approved by: flip1995
Pushing eceebc344887269a821d1df0c8693d9eb4d5da7d to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"flip1995","base_ref":"master","builders":{"checks-action_test":"https://github.com/rust-lang/rust-clippy/runs/1305406490","checks-action_dev_test":"https://github.com/rust-lang/rust-clippy/runs/1305369781","checks-action_remark_test":"https://github.com/rust-lang/rust-clippy/runs/1305367820"},"merge_sha":"eceebc344887269a821d1df0c8693d9eb4d5da7d"} -->
